### ISSUE TITLE
IDesignTimeDbContextFactory with UseModel produces incorrect model

### ISSUE THREAD

**Issue by hemirunner426** [dotnet/efcore#26802]
Consider this model:

```C#
public class TestModel
    {
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public long Id { get; set; }
        public string? Name { get; set; }
    }
```

Along with a design time DbContext factory:
```C#
internal class TestDbDesignContext : IDesignTimeDbContextFactory<TestContext>
    {
        public TestContext CreateDbContext(string[] args)
        {
            var optionsBuilder = new DbContextOptionsBuilder();
            optionsBuilder.UseSqlServer(@"Server=localhost;Database=TestDB;Trusted_Connection=True;");
            var modelBuilder = SqlServerConventionSetBuilder.CreateModelBuilder();
            modelBuilder.Entity<TestModel>();
            optionsBuilder.UseModel((IModel)modelBuilder.Model);
            return new TestContext(optionsBuilder.Options);
        }
    }
```

Running `dotnet ef migrations add Initial` produces the following migration:
```C#
public partial class Initial : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.CreateTable(
                name: "TestModel",
                columns: table => new
                {
                    Id = table.Column<long>(type: "bigint", nullable: false),
                    Name = table.Column<string>(type: "nvarchar(max)", nullable: true)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_TestModel", x => x.Id);
                });
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.DropTable(
                name: "TestModel");
        }
    }
```

In previous versions of EF core the ID colum would be appended like the following:
```C#
Id = table.Column<long>(nullable: false)
                        .Annotation("SqlServer:Identity", "1, 1"),
```

When the EF 6 migration is applied to the database the ID column is not set as `Identity = True`.  Thus, on insert the DbContext generates the following exception:

```
Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while saving the entity changes. See the inner exception for details.
 ---> Microsoft.Data.SqlClient.SqlException (0x80131904): Cannot insert the value NULL into column 'Id', table TestDB.dbo.TestModel'; column does not allow nulls. INSERT fails.
```

EF Core version: 6.0.0
Database provider:  Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6.0
Operating system: Windows 11
IDE: Visual Studio 2022

**Comment by ajcvickers** [comment]
@AndriySvyryd This seems to be caused by building the model externally from the context:

```C#
var modelBuilder = SqlServerConventionSetBuilder.CreateModelBuilder();
modelBuilder.Entity<TestModel>();
optionsBuilder.UseModel((IModel)modelBuilder.Model);
```

Explicitly finalizing the model doesn't help:

```C#
var modelBuilder = SqlServerConventionSetBuilder.CreateModelBuilder();
modelBuilder.Entity<TestModel>();
var finalizeModel = modelBuilder.FinalizeModel();
optionsBuilder.UseModel(finalizeModel);
```

@hemirunner426 As a workaround, configure the model in `OnModelCreating`.

**Comment by hemirunner426** [comment]
@ajcvickers Thanks for the suggestion I'll take a look to see if I can implement this workaround at design time

**Comment by AndriySvyryd** [comment]
This is a consequence of the [model initialization breaking change](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-6.0/breaking-changes#snapshot-initialization)

The full fix is:
```C# 
       public TestContext CreateDbContext(string[] args)
        {
            var optionsBuilder = new DbContextOptionsBuilder();
            optionsBuilder.UseSqlServer(@"Server=localhost;Database=TestDB;Trusted_Connection=True;");
            var modelBuilder = SqlServerConventionSetBuilder.CreateModelBuilder();
            modelBuilder.Entity<TestModel>();

            var model = modelBuilder.Model.FinalizeModel();
            var serviceContext = new TestContext(optionsBuilder.Options);
            model = serviceContext.GetService<IModelRuntimeInitializer>().Initialize(model);

            optionsBuilder.UseModel(model);

            return new TestContext(optionsBuilder.Options);
        }
```

We could add a check and throw an exception when this scenario is detected.

**Comment by hemirunner426** [comment]
> This is a consequence of the [model initialization breaking change](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-6.0/breaking-changes#snapshot-initialization)
> 
> The full fix is:
> 
> ```cs
>        public TestContext CreateDbContext(string[] args)
>         {
>             var optionsBuilder = new DbContextOptionsBuilder();
>             optionsBuilder.UseSqlServer(@"Server=localhost;Database=TestDB;Trusted_Connection=True;");
>             var modelBuilder = SqlServerConventionSetBuilder.CreateModelBuilder();
>             modelBuilder.Entity<TestModel>();
> 
>             var model = modelBuilder.Model.FinalizeModel();
>             var serviceContext = new TestContext(optionsBuilder.Options);
>             model = serviceContext.GetService<IModelRuntimeInitializer>().Initialize(model);
> 
>             optionsBuilder.UseModel(model);
> 
>             return new TestContext(optionsBuilder.Options);
>         }
> ```
> 
> We could add a check and throw an exception when this scenario is detected.

This workaround seemed to work.  Thanks.

**Comment by AndriySvyryd** [comment]
Fixing https://github.com/dotnet/efcore/issues/26186 would also fix this.

**Comment by ajcvickers** [comment]
Note from triage: fixing this is complicated/risky. For now, we will document how to create an appropriate model to pass to UseModel.

@AndriySvyryd  do we have a docs issue tracking documentation of this as a breaking change?

**Comment by AndriySvyryd** [comment]
@ajcvickers  It's already documented https://github.com/dotnet/EntityFramework.Docs/blob/main/entity-framework/core/what-is-new/ef-core-6.0/breaking-changes.md#snapshot-initialization
Just hasn't been pushed to live

**Comment by delikelli** [comment]
Hi! @AndriySvyryd . why do we have to create two instance from db context? another problem is we have a seperation with db context class and its dbcontextoptionsbuilder class and they are on different class libraries so we can not create object for dbcontext at method that uses model builder operations.

**Comment by AndriySvyryd** [comment]
@delikelli What is the reason for the dbcontextoptionsbuilder-related code to be in a different assembly? Perhaps there's a way of refactoring out the model building code so it can be also used for the design-time factory

**Comment by AndriySvyryd** [comment]
Fixed in [9e7002c](https://github.com/dotnet/efcore/commit/9e7002c9a93ac2717cdd394cf196278fed3a50eb)
